docs/doc/60-contributing/03-rfcs/20220704-presign.md

```diff
@@ -0,0 +1,179 @@
++--------+---------+---------------------------------------------------------------------------------+
+| GET    | []      | https://example.s3.amazonaws.com/books.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&... |
++--------+---------+---------------------------------------------------------------------------------+
+```
+
+By default, the presigned URL will be expired in 1 hour. Users can specify the expiry time to 2 hours like the following:
+
+```sql
+MySQL [(none)]> PRESIGN @my_stage/books.csv EXPIRE=7200
++--------+---------+---------------------------------------------------------------------------------+
+| method | headers | url                                                                             |
++--------+---------+---------------------------------------------------------------------------------+
+| GET    | []      | https://example.s3.amazonaws.com/books.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&... |
++--------+---------+---------------------------------------------------------------------------------+
+```
+
+A presigned URL is generated for the `read` action by default. Users can create a URL for `write` like the following:
+
+```sql
+MySQL [(none)]> PRESIGN @my_stage/books.csv EXPIRE=7200 ACTION=WRITE
++--------+---------+---------------------------------------------------------------------------------+
+| method | headers | url                                                                             |
++--------+---------+---------------------------------------------------------------------------------+
+| PUT    | []      | https://example.s3.amazonaws.com/books.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&... |
++--------+---------+---------------------------------------------------------------------------------+
+```
+
+If the `headers` returned by `presign` is not empty, users should include them in the actual requests.
+
+```sql
+MySQL [(none)]> PRESIGN @my_stage/books.csv ACTION=WRITE
++--------+--------------------------+---------------------------------------------------------------------------------+
+| method | headers                  | url                                                                             |
++--------+--------------------------+---------------------------------------------------------------------------------+
+| PUT    | [['x-amz-key', 'value']] | https://example.s3.amazonaws.com/books.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&... |
++--------+--------------------------+---------------------------------------------------------------------------------+
+```
+
+# Reference-level explanation
+
+`PRESIGN` will be implemented as a statement instead of a function so that we can return the HTTP method, headers, and URL at the same time.
+
+Most jobs have been done via [OpenDAL presign](https://opendal.databend.rs/rfcs/0413-presign.html).
+
+In databend, we will:
+
+- Add `PRESIGN` in the parser (only in the new planner)
```

Comment:
how about move the action part after the keyword `PRESIGN`? like:

```
PRESIGN UPLOAD @stage/file1.txt
PRESIGN DOWNLOAD @stage/file1.txt EXPIRE
PRESIGN UPLOAD_PART @stage/file1.txt
```

maybe we can extend the syntax like this in the future:

```
DOWNLOAD @stage/file1.txt
```

thus the `DOWNLOAD` statment and `PRESIGN DOWNLOAD` statement can be in the similar command style.

Reply:
That sounds a good idea!

Reply:
Addressed in https://github.com/datafuselabs/databend/pull/6503/commits/3802564a485f7ebb4d8251e2fabcd3a32da75401